tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,295 @@
+
+SubnetSendTransfer ==
+  (* Honestly sends a transfers cycles from one subnet to another. *)
+  numTransfers < MAX_TRANSFERS
+
+  \* Choose two subnets.
+  /\ \E sender, receiver \in SUBNETS: sender /= receiver
+
+    \* Choose an amount that's within the subnet's balance.
+    /\ \E amount \in 1..subnets[sender].balance:
+
+        \* Send a transfer to the receiver.
+        /\ subnetMsgs' = subnetMsgs \union {[
+                id |-> numTransfers,
+                type |-> TRANSFER,
+                from |-> sender,
+                to |-> receiver,
+                amount |-> amount
+            ]}
+
+        \* Subtract amount from sender's balance.
+        /\ subnets' = [subnets EXCEPT ![sender].balance = @ - amount]
+
+        \* Limit the number of transfers to keep the state space bounded.
+        /\ numTransfers' = numTransfers + 1
+
+        /\ UNCHANGED<<ledger, numDishonestActions>>
+
+SubnetDishonestSendTransfer ==
+    (* Maliciously transfers more cycles than the subnet's own balance. *)
+    numTransfers < MAX_TRANSFERS
+    /\ numDishonestActions < MAX_DISHONEST_TRANSFERS
+    /\ \E sender, receiver \in SUBNETS:
+        /\ sender /= receiver
+
+        \* Limit the number of dishonest actions to keep the state space bounded.
+        /\ numDishonestActions' = numDishonestActions + 1
+        /\ numTransfers' = numTransfers + 1
+
+        \* Choose an amount that's greater than the balance.
+        /\ \E counterfeit_cycles \in 1..MAX_COUNTERFEIT_CYCLES:
+            /\ LET amount == Max({subnets[sender].balance + counterfeit_cycles, 1}) IN
+                \* Send a transfer to the receiver.
+                /\ subnetMsgs' = subnetMsgs \union {[
+                        id |-> numTransfers,
+                        type |-> TRANSFER,
+                        from |-> sender,
+                        to |-> receiver,
+                        amount |-> amount
+                    ]}
+
+                \* Mark subnet as dishonest and subtract balance.
+                /\ subnets' = [subnets EXCEPT ![sender].honest = FALSE,
+                                              ![sender].balance = @ - amount]
+
+                /\ UNCHANGED<<ledger>>
+
+SubnetReceiveTransfer ==
+  \E msg \in subnetMsgs:
+      msg.type = TRANSFER
+      /\ subnets' = [
+                         \* Add amount in transfer to the unfinalized balance.
+          subnets EXCEPT ![msg.to].unfinalized[msg.from] = @ + msg.amount,
+                         \* Add message to ledger outgoing queue.
+                         ![msg.to].msgsToLedger = Append(@, msg)
+         ]
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = subnetMsgs \ {msg}
+      /\ UNCHANGED<<ledger, numDishonestActions, numTransfers>>
+
+LedgerPoll ==
+  \E s \in SUBNETS: Len(subnets[s].msgsToLedger) > 0
+    /\ LET msg == Head(subnets[s].msgsToLedger) IN
+      \* Send message to ledger.
+      /\ ledger' = [ledger EXCEPT !["msgs"][s] = Append(@,
```

Comment:
Actually, why do you need this transition? Couldn't the ledger read directly from `msgsToLedger`? I don't think you'd miss any relevant behaviors that way, but you'd have fewer transitions. I mean it doesn't hurt to have it from a soundness perspective, but it just seems to increase the state space needlessly.

Reply:
Also, nitpicking, but you could also use `!.msgs[s]` instead of `!["msgs"][s]` I think?

Reply:
I agree that it's not strictly required, but as I started working on getting rid of it, I felt that the drawbacks of having a slightly higher state space don't outweigh the gain of modeling the process more accurately in this case, so I'm thinking of leaving it as-is. If you feel strongly otherwise, let me know and I'd be happy to remove that transition.

Reply:
I don't feel strongly about it, but if you need to speed the model checking up further at some point, getting rid of `ledger.msgs` would be my suggestion.